src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,299 @@
+    # There should be exactly one grant request per registration request.
+    assert len(grant_requests) == len(registration_requests)
+    cbsd_ids = self.testcase.assertRegistered(registration_requests)
+
+    # Wrapped version of the grant requests
+    grant_requests_wrap = {
+      'grantRequest': grant_requests
+    }
+
+    #perform grant operation
+    grant_responses = self.testcase._sas.Grant(grant_requests_wrap,
+                                               self.ssl_cert, self.ssl_key)['grantResponse']
+
+    # Make one CBSD object per successful grant request
+    for index, grant_request in enumerate(grant_requests):
+      grant_request["cbsdId"] = cbsd_ids[index]
+      if grant_responses[index]['response']['responseCode'] == 0:
+        cbsd_object = Cbsd(cbsd_ids[index], registration_requests[index],
+                           [grant_responses[index]['grantId']], [grant_request])
+        self.cbsd_objects[cbsd_ids[index]] = cbsd_object
+
+  def getCbsdObjectById(self, cbsd_id):
+    try:
+      return self.cbsd_objects[cbsd_id]
+    except KeyError:
+      logging.debug("CBSD Object not found in the Domain Proxy object,cbsd_id=%s.", cbsd_id)
+
+  def heartbeatRequestForAllGrants(self):
+    """Performs heartbeat request for all grants of CBSD devices at once.
+
+    Returns:
+      heartbeat_requests : list of heart beat requests which are constructed for all the grants
+      by the caller(
+      heartbeat_responses: list of heart beat response
+    """
+    # Concatenate all the heartbeat request of CBSD into single request
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
+      heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+
+    if len(heartbeat_requests):
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      # Perform heartbeat requests
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_response in heartbeat_responses:
+        try:
+          cbsd_object = self.cbsd_objects[heartbeat_response['cbsdId']]
+          grant_object = cbsd_object.grant_objects[heartbeat_response['grantId']]
+        except KeyError:
+          logging.debug("CBSD or the Grant object not found")
+
+        if heartbeat_response['response']['responseCode'] == 0:
+          grant_object.authorized_in_last_heartbeat = True
+        else:
+          self.mapResponseCodeToGrantState(heartbeat_response['response']['responseCode'],
+                                           grant_object)
+      return heartbeat_requests, heartbeat_responses
+
+  def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state"""
+    if heartbeat_response_code == 502:
+      grant_object.is_terminated = True
```

Comment:
If an UNSYNC_OP_PARAM response code occurs it's probably worth at least logging an error. Removing this check and implement a check at the bottom for any non-matching responses (as comment on line 221 suggests) should be sufficient.

Reply:
Error   logged in case of  Response code   UNSYNC_OP_PARAM